compiler/AST/foralls.cpp

```diff
@@ -1216,3 +1226,78 @@ void resolveForallStmts2() {
     convertIteratorForLoopexpr(fs);
   }
 }
+
+
+///////////////////////////////
+//                           //
+//   PRIM_REDUCE lowering    //
+//                           //
+///////////////////////////////
+
+// Insert a call temp. It is simpler than the full-blown normalize()
+// and the caller can reference the temp.
+static SymExpr* normalizeIITR(Expr* ref, Expr* iitr) {
+  VarSymbol* temp = newTemp("iitr_temp");
+  temp->addFlag(FLAG_TYPE_VARIABLE);
+  ref->insertBefore(new DefExpr(temp));
+  ref->insertBefore("'move'(%S,%E)", temp, iitr);
+  return new SymExpr(temp);
+}
+
+// Given a reduce expression like "op reduce data", return op(inputType).
+// inputType is the type of things being reduced - when iterating over 'data'.
+// This matches the case where inputType is provided by the user.
+static Expr* lowerReduceOp(Expr* ref, SymExpr* opSE, SymExpr* dataSE,
+                           bool zippered)
+{
+  CallExpr* iit = NULL;
+  if (zippered) {
+    // Cf. destructZipperedIterables. 'zipcall' will be removed there.
+    CallExpr* zipcall = toCallExpr(getDefOfTemp(dataSE)->copy());
+    INT_ASSERT(zipcall->isPrimitive(PRIM_ZIP));
+    iit = new CallExpr("iteratorIndexTypeZip");
+    for_actuals(actual, zipcall)
+      iit->insertAtTail(toSymExpr(actual)->symbol());
+  } else {
+    iit = new CallExpr("iteratorIndexType", dataSE->symbol());
+  }
+
+  ref->insertBefore(iit);
+  Expr* iitR = resolveExpr(iit)->remove();
+  if (!isSymExpr(iitR)) iitR = normalizeIITR(ref, iitR);
+
+  return new CallExpr(opSE, iitR);
+}
+
+// Within resolveBlockStmt / for_exprs_postorder framework,
+// we need to lower PRIM_REDUCE prior to the resolveCall()
+// that gets invoked from resolveExpr(). That way the
+// ForallStmt plus scaffolding that lowerPrimReduce() injects
+// can come after 'retval'. resolveCall() does not support that.
+
+void lowerPrimReduce(CallExpr* call, Expr*& retval) {
+  if (call->id == breakOnResolveID) gdbShouldBreakHere();
+
+  Expr*   callStmt = call->getStmtExpr();
+  CallExpr*   noop = new CallExpr(PRIM_NOOP);
+  callStmt->insertBefore(noop);
+
+  SymExpr*   opSE = toSymExpr(call->get(1)->remove());           // 1st arg
+  SymExpr* dataSE = toSymExpr(call->get(1)->remove());           // 2nd arg
+  bool   zippered = toSymExpr(call->get(1))->symbol() == gTrue;  // 3rd arg
+  bool  reqSerial = false; // vass todo - require in some cases
```

Comment:
looks like something to address before merging. Maybe the comment can simply be deleted? (I don't see why a reduce expression would require serial iteration). At the very least update the comment and make an issue.

Reply:
I deferred this to #11819.